### DISCUSSION TITLE
Performance comparison of the most popular tween libraries for Unity.

### DISCUSSION THREAD

**Discussion by KyryloKuzyk** [KyryloKuzyk/PrimeTween#10] [Announcements]
Performance comparison of the most popular tween libraries for Unity. 
---  

All benchmarks in this article are performed with the official [Performance testing API](https://docs.unity3d.com/Packages/com.unity.test-framework.performance@3.0/manual/index.html) and can be easily replicated by cloning the [Benchmarks](https://github.com/KyryloKuzyk/PrimeTween/tree/main/Benchmarks) repo.  For a detailed comparison of PrimeTween vs. DOTween, please visit the dedicated [article](https://github.com/KyryloKuzyk/PrimeTween/discussions/8).

### Runtime Performance  
> Test environment: Unity 2022.3.9, Macbook Pro M1 32GB, IL2CPP build, 100.000 iterations per test, average frame time in milliseconds (lower numbers are better).  
> 
|                              | [DOTween 1.2.745](https://github.com/KyryloKuzyk/PrimeTween/blob/main/Benchmarks/Assets/DOTween_PrimeTweenTests.cs) | [LeanTween 2.51](https://github.com/KyryloKuzyk/PrimeTween/blob/main/Benchmarks/Assets/LeanTweenTests.cs) | [UnityTweens 3.2.0](https://github.com/KyryloKuzyk/PrimeTween/blob/main/Benchmarks/Assets/UnityTweensTests.cs) | [PrimeTween 1.0.10](https://github.com/KyryloKuzyk/PrimeTween/blob/main/Benchmarks/Assets/DOTween_PrimeTweenTests.cs) | PrimeTween vs second best |
|------------------------------|-----------------|----------------|-------------------|-------------------|---------------------------|
| Animation start              |        33.54 ms |       15.00 ms |          43.18 ms |           5.76 ms |                  **2.6x** |
| Animation start (all params) |        38.01 ms |       24.68 ms |          39.73 ms |           6.72 ms |                 **3.67x** |
| Position animation           |         8.91 ms |       12.54 ms |           7.88 ms |           4.34 ms |                 **1.82x** |
| Rotation animation           |         9.74 ms |       16.28 ms |           9.49 ms |           6.79 ms |                  **1.4x** |
| Custom animation             |         4.93 ms |        4.45 ms |           4.78 ms |           3.28 ms |                 **1.36x** |
| Custom animation curve       |        12.11 ms |        9.90 ms |          14.01 ms |           4.84 ms |                 **2.04x** |
| Delay                        |         4.17 ms |        4.26 ms |           1.99 ms |           2.07 ms |                 **0.96x** |
| Delay start                  |        15.37 ms |       13.07 ms |          38.52 ms |           7.32 ms |                 **1.79x** |
| Sequence with 3 tweens       |         9.36 ms |        9.49 ms |                 - |           2.83 ms |                 **3.31x** |
| Sequence start               |        45.59 ms |    49963.62 ms |                 - |           8.83 ms |                 **5.16x** |

### GC Memory Allocations  
> The amount of GCAlloc per **one iteration**, lower numbers are better.  

|           | DOTween 1.2.745 | LeanTween 2.51 | UnityTweens 3.2.0 | PrimeTween 1.0.10 |
|-----------|-----------------|----------------|-------------------|-------------------|
| Animation | 734 B           |          292 B |             878 B |           **0 B** |
| Delay     | 584 B           |          146 B |             944 B |           **0 B** |
| Sequence  | 2846 B          |          877 B |                 - |           **0 B** |

### How to run the benchmarks?
After cloning the [Benchmarks](https://github.com/KyryloKuzyk/PrimeTween/tree/main/Benchmarks) repo, open the '*Window/General/**Test Runner***' and run the tests on the target platform by pressing the '*Run All Tests (TargetPlatform)*' button.

This repo has the `PRIME_TWEEN_DISABLE_ASSERTIONS` define. PrimeTween heavily uses assertions in development builds, and this define will disable them to match the release performance in tests.

Noticed an error in the measurement method? Or have something to add? You are welcome to join the discussion.

**Comment by yosun** [comment]
niiice! but also curious why yours is slower on delay? 

also typo in leantween sequence start?

**Reply by KyryloKuzyk** [reply]
> why yours is slower on delay?

I assume this is because LeanTween has a special treatment of delays, while in PrimeTween delay is just a regular tween with that doesn't animate anything. Writing a custom delay optimization wasn't worth the benefit in my opinion.

> typo in leantween sequence start?

Not a typo, LeanTween has O(n^2) complexity in starting sequences. It may be unnoticeable if your project uses a couple of hundreds of sequences, but you should keep the exponential complexity in mind.

**Comment by fslse** [comment]
About the GC Memory Allocations of DOTween, may I ask if you have enabled the tween recycling option?

**Reply by KyryloKuzyk** [reply]
> About the GC Memory Allocations of DOTween, may I ask if you have enabled the tween recycling option?

Yes, of course! Please see the 'GC Memory Allocations' in this article for more details on the testing methodology around the memory allocation: https://github.com/KyryloKuzyk/PrimeTween/discussions/8

In short, I tested two most common scenarios: 1) creating the tween reference in OnKill(), and 2) not storing the tween reference at all. The 2) option still allocates 324 B of GC every tween start.

**Comment by EnterUserNamePlease** [comment]
Hi!
We already have a working prototype of a mobile game where DOTween is used very frequently. I was aware of PrimeTween and planned to replace DOTween with it at the end. Today, I replaced a couple of placements in the game and started profiling to confirm the improvements. But to my surprise, the profiler showed numbers almost twice as bad - operations took twice as long as with DOTween. This is not what I expected! 🙂

To be more specific, I'm moving around 30-40 Transforms with onUpdate and onComplete. I started diving into forums and found a benchmark. I ran the test, and the report also showed that PrimeTween performs worse than DOTween.

Why is this happening, and is it expected behavior? Performance is the main reason for switching, but it turns out that it's not working as intended.
Two reports from two different laptops:
![image](https://github.com/user-attachments/assets/8fa94409-7a27-48ba-a3e5-aa1b2a792fc5)
![image](https://github.com/user-attachments/assets/dc07342d-94fa-4981-9a2e-9c4c6f4da586)

**Reply by KyryloKuzyk** [reply]
Hey, @EnterUserNamePlease, sorry for the late reply.

Hmm, these numbers indeed look strange to me. Can you please check that you have the `PRIME_TWEEN_DISABLE_ASSERTIONS` define  in 'Player Settings / Other Settings / Script Compilation / Scripting Define Symbols'? With this define PrimeTween will run at release performance.

I just did these tests on different platforms and the results differ by a lot.
MacBook Pro M1 (Apple silicon): **1.88x**
<img width="358" alt="image" src="https://github.com/user-attachments/assets/716fd365-8264-497d-b1a5-a84920e57d13" />

Google Pixel 3a (200000 iterations instead of 100000): **0.9x**
<img width="382" alt="image" src="https://github.com/user-attachments/assets/4101bbfb-a44f-489f-88b1-70947f093c6a" />

iPhone 13 Pro (500000 iterations instead of 100000): **1.43x**
<img width="376" alt="image" src="https://github.com/user-attachments/assets/f5b04d7e-35fa-4055-aa37-e57c4dd6fb8d" />

PrimeTween performs worse than DOTween on the old Android device, but better on the iPhone and MacBook. This requires a deeper investigation. Probably, I have to write tests that measure the main update loop of both tweening libraries without the "com.unity.test-framework.performance" package.

Can you please tell me the models of your laptops?

> To be more specific, I'm moving around 30-40 Transforms with onUpdate and onComplete. I started diving into forums and found a benchmark. I ran the test, and the report also showed that PrimeTween performs worse than DOTween.

Saying all that, the synthetic tests in the repo run with 100000 simultaneous tweens, but 30-40 tweens should be impossible to measure precisely. While these synthetic tests on your machines show that PrimeTween is **1.29x** slower than DOTween, PrimeTween still doesn't allocate any garbage, which in real-world scenarios is very important to maintain stable framerates.

**Reply by EnterUserNamePlease** [reply]
Hi, thanks for response!
Laptop models: Lenovo Slim 7 16IAH7 (i7-12700H) + Lenovo Legion 7 16IRX9 (i9-14900HX).
I made more tests with and without `DISABLE_ASSERTIONS` define on Slim 7:

<img width="1115" alt="Screenshot_2" src="https://github.com/user-attachments/assets/97fa3a42-1605-41a6-b231-a022e6c250ae" />
<img width="1115" alt="Screenshot_3" src="https://github.com/user-attachments/assets/52fa4d47-955a-45c3-af9f-ace4bf2a4944" />

I want to launch the benchmark on my android devices, can you please give the step-by-step instructions on how to do it?

**Reply by KyryloKuzyk** [reply]
@EnterUserNamePlease Oh right, running performance tests in the Editor is not representative, you should measure a real build instead. To run the tests on an Android device, select the tests you want to run, then press a dropdown arrow near the "Run All Tests (Android)" button, then press "Run Selected Tests".

**Reply by EnterUserNamePlease** [reply]
Hi again, I am back with real build tests!)

Galaxy Note10+ (SM-N975F/DS)
<img width="695" alt="note 10+" src="https://github.com/user-attachments/assets/d8aad67c-296a-4459-ac42-e4f250eb09c0" />

Galaxy A12 (SM-A125F/DSN)
<img width="690" alt="A12" src="https://github.com/user-attachments/assets/14f7da82-6cbe-49ca-848b-1955755095d6" />

Meizu M5 Note (yeah, today it can be equated to a calculator, but it was still interesting to see the results))
<img width="693" alt="m5 note" src="https://github.com/user-attachments/assets/f6bb26e8-73ef-41d8-ad9d-4e63bcd84fd7" />

If I understand correctly, PrimeTween can be a little slower sometimes, but it definitely doesn't allocate garbage, which is more important.

**Reply by KyryloKuzyk** [reply]
@EnterUserNamePlease  I addressed a minor measurement issue in https://github.com/KyryloKuzyk/PrimeTween/commit/3b27995f5bd02e9398068836b4c750f49c28b4c7. As you can see in the image you shared, the measurement of 200 frames can take more time than the actual animation on Android devices. I fixed that, so measuring mobile devices became better now.

**Comment by RobertJaskowski** [comment]
Please include LitMotion comparison for non obfuscated data.

**Comment by PoyrazGoksel** [comment]
You are a genius, a perfect asset, and there is no error hiding, either. Yum!
